File: backend/app/routers/workflow.py
from fastapi import APIRouter, Depends
from app.services import workflow
from app.services.session import require_permission
from app.types.workflow import SaveGraphRequest, ToggleWorkflowRequest

router = APIRouter()

@router.get("")
async def list_workflows(_=Depends(require_permission("Workflow.Read"))):
	return await workflow.list_workflows()

@router.get("/{document_type}")
async def get_workflow(document_type: str, _=Depends(require_permission("Workflow.Read"))):
	return await workflow.get_workflow(document_type)

@router.patch("/{document_type}")
async def toggle_workflow(document_type: str, data: ToggleWorkflowRequest, _=Depends(require_permission("Workflow.Write"))):
	return await workflow.toggle_workflow(document_type, data.is_active)

@router.put("/{document_type}/graph")
async def save_graph(document_type: str, data: SaveGraphRequest, _=Depends(require_permission("Workflow.Write"))):
	return await workflow.save_graph(document_type, data)

@router.get("/{document_type}/{record_id}/history")
async def get_workflow_history(document_type: str, record_id: int, _=Depends(require_permission())):
	return await workflow.get_workflow_history(document_type, record_id)

@router.post("/{document_type}/{record_id}/submit")
async def submit_workflow(document_type: str, record_id: int, _=Depends(require_permission())):
	return await workflow.submit_workflow(document_type, record_id)

@router.post("/{document_type}/{record_id}/approve")
async def approve_workflow(document_type: str, record_id: int, _=Depends(require_permission())):
	return await workflow.approve_workflow(document_type, record_id)

@router.post("/{document_type}/{record_id}/reject")
async def reject_workflow(document_type: str, record_id: int, _=Depends(require_permission())):
	return await workflow.reject_workflow(document_type, record_id)

File: frontend/app/lib/api/workflow.ts
import { getJSON, patchJSON, postJSON } from "./submissions";
import { APIResult } from "../types/data";
import { WorkflowSummary, WorkflowDetail, SaveGraphRequest } from "../types/workflow";
import { API_BASE_URL } from "../config";

export async function listWorkflows(): Promise<APIResult<WorkflowSummary[]>> {
	return getJSON("/workflows");
}

export async function getWorkflow(documentType: string): Promise<APIResult<WorkflowDetail>> {
	return getJSON(`/workflows/${documentType}`);
}

export async function toggleWorkflow(documentType: string, isActive: boolean): Promise<APIResult<WorkflowSummary>> {
	return patchJSON(`/workflows/${documentType}`, { is_active: isActive });
}

export async function submitWorkflow(documentType: string, recordId: number): Promise<APIResult<void>> {
	return postJSON(`/workflows/${documentType}/${recordId}/submit`);
}

export async function approveWorkflow(documentType: string, recordId: number): Promise<APIResult<void>> {
	return postJSON(`/workflows/${documentType}/${recordId}/approve`);
}

export async function rejectWorkflow(documentType: string, recordId: number): Promise<APIResult<void>> {
	return postJSON(`/workflows/${documentType}/${recordId}/reject`);
}

export async function saveGraph(documentType: string, data: SaveGraphRequest): Promise<APIResult<void>> {
	const res = await fetch(`${API_BASE_URL}/workflows/${documentType}/graph`, {
		method: "PUT",
		headers: { "Content-Type": "application/json" },
		credentials: "include",
		body: JSON.stringify(data),
	});
	const ct = res.headers.get("content-type") || "";

	return ct.includes("application/json") ? res.json() : (res.text() as unknown as Promise<any>);
}
